src/main/java/com/android/volley/cronet/CronetHttpStack.java

```diff
@@ -0,0 +1,248 @@
+                            UrlRequest urlRequest, UrlResponseInfo urlResponseInfo, String s) {
+                        urlRequest.followRedirect();
+                    }
+
+                    @Override
+                    public void onResponseStarted(
+                            UrlRequest urlRequest, UrlResponseInfo urlResponseInfo) {
+                        int httpStatusCode = urlResponseInfo.getHttpStatusCode();
+                        // what checks should we be making of the status code, if any?
+                        if (mPool != null) {
+                            urlRequest.read(ByteBuffer.wrap(mPool.getBuf(102400)));
+                        } else {
+                            urlRequest.read(ByteBuffer.allocateDirect(102400));
+                        }
+                    }
+
+                    @Override
+                    public void onReadCompleted(
+                            UrlRequest urlRequest,
+                            UrlResponseInfo urlResponseInfo,
+                            ByteBuffer byteBuffer) {
+                        byteBuffer.flip();
+                        try {
+                            receiveChannel.write(byteBuffer);
+                            byteBuffer.clear();
+                            urlRequest.read(byteBuffer);
+                        } catch (IOException e) {
+                            callback.onError(e);
+                            urlRequest.cancel();
+                        }
+                    }
+
+                    @Override
+                    public void onSucceeded(
+                            UrlRequest urlRequest, UrlResponseInfo urlResponseInfo) {
+                        List<Header> headers = getHeaders(urlResponseInfo.getAllHeadersAsList());
+                        HttpResponse response =
+                                new HttpResponse(
+                                        urlResponseInfo.getHttpStatusCode(),
+                                        headers,
+                                        bytesReceived.toByteArray());
+                        callback.onSuccess(response);
+                    }
+
+                    @Override
+                    public void onFailed(
+                            UrlRequest urlRequest,
+                            UrlResponseInfo urlResponseInfo,
+                            CronetException e) {
+                        if (urlResponseInfo != null) {
+                            List<Header> headers =
+                                    getHeaders(urlResponseInfo.getAllHeadersAsList());
+                            HttpResponse response =
+                                    new HttpResponse(
+                                            urlResponseInfo.getHttpStatusCode(),
+                                            headers,
+                                            bytesReceived.toByteArray());
+                            callback.onSuccess(response);
```

Comment:
yeah, I think you're right. I was thinking that any information should be sent over, but if it's only partial or has other errors, it's probably better to call onError. And I think regardless of the response code, as long as the read is successful, it will end up in onSucceeded, not onFailed.